x-pack/platform/packages/shared/kbn-streamlang/src/validation/validate_streamlang.ts

```diff
@@ -0,0 +1,702 @@
+import { flattenSteps } from '../transpilers/shared/flatten_steps';
+import { isAlwaysCondition } from '../../types/conditions';
+import { parseGrokPattern, parseDissectPattern } from '../../types/utils';
+
+/**
+ * Supported field types for type tracking
+ */
+export type FieldType = 'string' | 'number' | 'boolean' | 'date' | 'unknown';
+
+/**
+ * Map of field names to their inferred types
+ */
+export type FieldTypeMap = Map<string, FieldType>;
+
+export interface StreamlangValidationError {
+  type: 'non_namespaced_field' | 'reserved_field' | 'type_mismatch' | 'mixed_type';
+  message: string;
+  processorId?: string;
+  field: string;
+  expectedType?: FieldType | FieldType[];
+  actualType?: FieldType;
+  conflictingTypes?: FieldType[];
+}
+
+/**
+ * Human-readable labels for validation error types
+ */
+export const validationErrorTypeLabels = {
+  non_namespaced_field: i18n.translate('xpack.streamlang.validation.nonNamespacedField', {
+    defaultMessage: 'Non-namespaced field',
+  }),
+  reserved_field: i18n.translate('xpack.streamlang.validation.reservedField', {
+    defaultMessage: 'Reserved field',
+  }),
+  type_mismatch: i18n.translate('xpack.streamlang.validation.typeMismatch', {
+    defaultMessage: 'Type mismatch',
+  }),
+  mixed_type: i18n.translate('xpack.streamlang.validation.mixedType', {
+    defaultMessage: 'Mixed type',
+  }),
+};
+
+export interface StreamlangValidationOptions {
+  /**
+   * Whether to validate for wired streams (requires namespaced fields)
+   */
+  isWiredStream?: boolean;
+  /**
+   * List of reserved/forbidden field names that cannot be modified by processors
+   */
+  reservedFields?: string[];
+  /**
+   * Whether to perform type checking on field usage
+   */
+  validateTypes?: boolean;
+  /**
+   * Initial field types (e.g., from parent stream or existing fields)
+   */
+  initialFieldTypes?: FieldTypeMap;
+}
+
+export interface StreamlangValidationResult {
+  isValid: boolean;
+  errors: StreamlangValidationError[];
+  /**
+   * Field types available at each processor (indexed by processorId).
+   * This represents the field types that exist BEFORE the processor runs,
+   * which is useful for field selection in the UI.
+   */
+  fieldTypesByProcessor: Map<string, FieldTypeMap>;
+}
+
+/**
+ * Check if a field is a namespaced ECS field or an allowed keep field.
+ * Based on the logic from @kbn/streams-schema/src/helpers/namespaced_ecs.ts
+ *
+ * Namespaced ECS fields follow the pattern: namespace.field or namespace.nested.field
+ * Examples: attributes.custom, body.structured.data, resource.attributes.host
+ *
+ * Keep fields are special fields that are allowed without namespacing:
+ * @timestamp, trace_id, span_id, severity_text, body, severity_number, event_name, etc.
+ */
+function isNamespacedEcsField(fieldName: string): boolean {
+  // List of special fields that are allowed without namespacing (from kbn-streams-schema)
+  const keepFields = [
```

Comment:
Should `keepFields` and `namespacePrefixes` be exported so that `@kbn/streams-schema` can re-export them from `@kbn/streamlang` at:

https://github.com/elastic/kibana/blob/e84e2ebcadd9232e4da97f4a9ac8c5a66212a4ff/x-pack/platform/packages/shared/kbn-streams-schema/src/helpers/namespaced_ecs.ts#L8

Reply:
good idea, did that